### PR TITLE
Add keycodes file and replace keycodes accordingly

### PR DESCRIPTION
**Pull request by zMoooooritz**
This does implement the changes we discussed in #27 concerning the keycodes action.
As always feel free to change things if required!

### CODE REVIEW

File: keycodes.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
+package main
+
+import (
+	"strconv"
+	"strings"
+)
+
+func formatKeycodes(keycode string) string {
+	kc := map[string]int{
```
Comment (muesli) [inline keycodes.go:9]: This is probably a map we only want initialized once, so it doesn't have to be allocated and freed for every call of this function. I'm fine with using global vars in such cases.

File: keycodes.go [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+		"KeyBrightnessZero":   244,
+		"KeyDisplayOff":       245,
+		"KeyWimax":            246,
+		"KeyRfkill":           247,
+		"KeyMicmute":          248,
+		"keyMax":              248,
+	}
+
+	for k, v := range kc {
+		keycode = strings.ReplaceAll(keycode, k, strconv.Itoa(v))
```
Comment (muesli) [inline keycodes.go:257]: We should call `strings.ToLower` on both `keycode` and `k` here, so we match any capitalization of the string.

File: keycodes.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
+package main
+
+import (
+	"strconv"
+	"strings"
+)
+
+var kc = map[string]int{
+	"keyReserved":         0,
```
Comment (muesli) [inline keycodes.go:9]: We can drop this one, it only makes sense in the context of uinput.

File: keycodes.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	"KeyUnknown":          240,
+	"KeyVideoNext":        241,
+	"KeyVideoPrev":        242,
+	"KeyBrightnessCycle":  243,
+	"KeyBrightnessZero":   244,
+	"KeyDisplayOff":       245,
+	"KeyWimax":            246,
+	"KeyRfkill":           247,
+	"KeyMicmute":          248,
+	"keyMax":              248,
```
Comment (muesli) [inline keycodes.go:252]: We can drop this one, it only makes sense in the context of uinput.

**Comment by muesli** [comment]
One more thing I'm wondering about: all the items in the map have the prefix `Key`. We could probably just drop that prefix altogether?

**Comment by zMoooooritz** [comment]
> One more thing I'm wondering about: all the items in the map have the prefix `Key`. We could probably just drop that prefix altogether?

Have not thought about that one until now. But yes this would make it even more intuitive and convenient to set-up 'macros'

**Comment by zMoooooritz** [comment]
Note: I have changed int to uint8 as this should be a bit more memory efficient and the map should not change in the near future

**Comment by muesli** [comment]
Last but not least: we should add a reference to `keycodes.go` in the README PR #27 :smiley:

**Review by muesli — APPROVED** [review]

**Comment by zMoooooritz** [comment]
> Last but not least: we should add a reference to `keycodes.go` in the README PR #27

Yes of course. I wanted to wait until the changes in this branch are final and then adapt the README accordingly

**Comment by muesli** [comment]
Heads up: I just realized an issue with the latest implementation. Something like `VolumeDown` would cause a series of keycode replacements, as literally every single character in this string matches a keycode in the map. Fixing this right now.

**Comment by zMoooooritz** [comment]
> Heads up: I just realized an issue with the latest implementation. Something like `VolumeDown` would cause a series of keycode replacements, as literally every single character in this string matches a keycode in the map. Fixing this right now.
> 
> In hindsight: this wouldn't have happened with a `Key` prefix

It's too late that is something I have to catch sry :/
Maybe we can keep the map as it is and split the input string at `-` and try to index with each substring into the map?

**Comment by muesli** [comment]
> It's too late that is something I have to catch sry :/

No worries, I should have realized this during my review as well.

> Maybe we can keep the map as it is and split the input string at `-` (watch out for the special case of `+` when there is also a delay) and try to index with each substring into the map?

Just pushed a fix to master doing pretty much that! I also realized the numbers (1, 2, 3 etc.) clash with the literal keycodes, so I gave them a `Num` prefix.

Final thought: we could be cleverer about using the map and do actual lookups. We would have to store all entries in lowercase tho. Not that it would make a huge difference as this only gets called on button presses and the map is fairly tiny.

**Pull request merged**